core/domain/email_manager_test.py

```diff
@@ -426,3 +428,422 @@ def test_record_of_sent_email_is_written_to_datastore(self):
+            self.assertEqual(0, len(messages))
+
+            # Check that the content of this email was not recorded in
+            # SentEmailModel.
+            all_models = email_models.SentEmailModel.get_all().fetch()
+            self.assertEqual(len(all_models), 1)
+
+    def test_send_email_does_not_resend_within_duplicate_interval(self):
+        can_send_emails_ctx = self.swap(
+            feconf, 'CAN_SEND_EMAILS_TO_USERS', True)
+
+        duplicate_email_ctx = self.swap(
+            feconf, 'DUPLICATE_EMAIL_INTERVAL_MINS', 2)
+
+        logged_errors = []
+
+        def _log_error_for_tests(error_message):
+            logged_errors.append(error_message)
+
+        log_new_error_counter = test_utils.CallCounter(_log_error_for_tests)
+        log_new_error_ctx = self.swap(
+            email_manager, 'log_new_error', log_new_error_counter)
+
+        with can_send_emails_ctx, duplicate_email_ctx, log_new_error_ctx:
+            config_services.set_property(
+                self.admin_id, email_manager.EMAIL_SENDER_NAME.name,
+                'Email Sender')
+
+            all_models = email_models.SentEmailModel.get_all().fetch()
+            self.assertEqual(len(all_models), 0)
+
+            # pylint: disable=protected-access
+            email_manager._send_email(self.new_user_id,
+                                      feconf.SYSTEM_COMMITTER_ID,
+                                      feconf.EMAIL_INTENT_SIGNUP,
+                                      'Email Subject', 'Email Body')
+
+            # Check that a new email was sent.
+            messages = self.mail_stub.get_sent_messages(
+                to=self.NEW_USER_EMAIL)
+            self.assertEqual(1, len(messages))
+
+            # Check that the content of this email was recorded in
+            # SentEmailModel.
+            all_models = email_models.SentEmailModel.get_all().fetch()
+            self.assertEqual(len(all_models), 1)
+
+            email_manager._send_email(self.new_user_id,
+                                      feconf.SYSTEM_COMMITTER_ID,
+                                      feconf.EMAIL_INTENT_SIGNUP,
+                                      'Email Subject', 'Email Body')
+            # pylint: enable=protected-access
+
+            # Two errors should be recorded in the logs.
+            self.assertEqual(log_new_error_counter.times_called, 2)
+            self.assertRegexpMatches(logged_errors[0], 'Duplicate email')
+
+            # Check that a new email was not sent.
+            messages = self.mail_stub.get_sent_messages(
+                to=self.NEW_USER_EMAIL)
+            self.assertEqual(1, len(messages))
+
+            # Check that the content of this email was not recorded in
+            # SentEmailModel.
+            all_models = email_models.SentEmailModel.get_all().fetch()
+            self.assertEqual(len(all_models), 1)
+
+    def test_send_email_sends_within_duplicate_interval_diff_recipient(self):
```

Comment:
Hard to understand. Maybe: test_sending_duplicate_email_with_different_recipient(self):